src/back/hlsl/writer.rs

```diff
@@ -72,42 +92,267 @@ impl<W: Write> Writer<W> {
+                named_expressions: &function.named_expressions,
             };
-            if ep.stage == ShaderStage::Compute {
-                // HLSL is calling workgroup size, num threads
-                let num_threads = ep.workgroup_size;
-                writeln!(
-                    self.out,
-                    "[numthreads({}, {}, {})]",
-                    num_threads[0], num_threads[1], num_threads[2]
-                )?;
+            let name = self.names[&NameKey::Function(handle)].clone();
+
+            self.write_function(module, name.as_str(), function, &ctx)?;
+
+            writeln!(self.out)?;
+        }
+
+        // Write entry point
+        let ep = &module.entry_points[ep_idx];
+        self.write_ep_input_struct(module, &ep.function, ep.stage)?;
+
+        let func_ctx = FunctionCtx {
+            ty: FunctionType::EntryPoint(ep_idx as u16),
+            info: info.get_entry_point(ep_idx),
+            expressions: &ep.function.expressions,
+            named_expressions: &ep.function.named_expressions,
+        };
+
+        if ep.stage == ShaderStage::Compute {
+            // HLSL is calling workgroup size, num threads
+            let num_threads = ep.workgroup_size;
+            writeln!(
+                self.out,
+                "[numthreads({}, {}, {})]",
+                num_threads[0], num_threads[1], num_threads[2]
+            )?;
+        }
+
+        // We write 1 ep per stage. Name already unique
+        self.write_function(module, &ep.name, &ep.function, &func_ctx)?;
+
+        Ok(())
+    }
+
+    fn write_binding(&mut self, binding: &crate::Binding) -> BackendResult {
+        match *binding {
+            crate::Binding::BuiltIn(builtin) => {
+                write!(self.out, " : {}", builtin_str(builtin))?;
+            }
+            crate::Binding::Location { location, .. } => {
+                write!(self.out, " : {}{}", LOCATION_SEMANTIC, location)?;
             }
-            self.write_function(module, &ep.function, &func_ctx)?;
+        }
 
-            if index < module.entry_points.len() - 1 {
+        Ok(())
+    }
+
+    fn write_ep_input_struct(
+        &mut self,
+        module: &Module,
+        func: &Function,
+        stage: ShaderStage,
+    ) -> BackendResult {
+        if !func.arguments.is_empty() {
+            let struct_name = match stage {
+                ShaderStage::Vertex => self.namer.call_unique("VertexInput"),
+                ShaderStage::Fragment => self.namer.call_unique("FragmentInput"),
+                ShaderStage::Compute => self.namer.call_unique("ComputeInput"),
+            };
+
+            let mut members = Vec::with_capacity(func.arguments.len());
+
+            write!(self.out, "struct {}", &struct_name)?;
+            writeln!(self.out, " {{")?;
+
+            for arg in func.arguments.iter() {
+                let arg_name = if let Some(ref name) = arg.name {
+                    name
+                } else {
+                    "member"
```

Comment:
Maybe "argument"? I'd expect to see "member" for the structure members but maybe not in the arguments

Reply:
That's not a good variable name. It should be `member_name` since all entry point arguments are wrapped into a struct. All function arguments for an entry point function become `wrappeed_struct.member`.